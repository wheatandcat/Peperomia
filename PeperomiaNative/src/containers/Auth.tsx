import * as GoogleSignIn from "expo-google-sign-in";
import * as Google from "expo-google-app-auth";
import { AsyncStorage, Platform } from "react-native";
import React, { createContext, Component } from "react";
import Constants from "expo-constants";
import { Sentry } from "react-native-sentry";
import * as firebase from "firebase";

const Context = createContext({});
const { Provider } = Context;

interface Props {}

interface State {
  uid: string;
  email: string;
}

const isStandaloneAndAndroid = () => {
  return Platform.OS === "android" && Constants.appOwnership !== "expo";
};

export default class extends Component<Props, State> {
  state = {
    email: "",
    uid: ""
  };

  async componentDidMount() {
    if (isStandaloneAndAndroid()) {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      try {
        await GoogleSignIn.initAsync({
          clientId: String(androidClientId)
        });
      } catch ({ message }) {
        Sentry.captureMessage(JSON.stringify(message));
      }
    }

    const loggedIn = await this.loggedIn();

    if (loggedIn && !this.state.uid) {
      const uid = await AsyncStorage.getItem("uid");
      if (uid) {
        this.setState({
          uid
        });
      }
    }

    if (loggedIn && !this.state.email) {
      const email = await AsyncStorage.getItem("email");
      if (email) {
        this.setState({
          email
        });
      }
    }
  }

  onGoogleLogin = async () => {
    if (isStandaloneAndAndroid()) {
      // TODO: AndroidのstandaloneのみGoogleSignInを使わないとエラーになる
      // https://github.com/expo/expo/issues/4762

      await GoogleSignIn.askForPlayServicesAsync();
      const result = await GoogleSignIn.signInAsync();

      if (result.type === "success" && result.user && result.user.auth) {
        const { idToken, accessToken } = result.user.auth;
        await this.firebaseLogin(idToken || "", accessToken || "");
      } else {
        Sentry.captureMessage(JSON.stringify(result));
      }
    } else {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      const iosClientId = process.env.GOOGLE_LOGIN_IOS_CLIENT_ID;
      const result = await Google.logInAsync({
        clientId:
          Platform.OS === "ios" ? String(iosClientId) : String(androidClientId),
        iosClientId,
        androidClientId,
        scopes: ["profile", "email"]
      });

      if (result.type === "success") {
        const { idToken, accessToken } = result;
        await this.firebaseLogin(idToken || "", accessToken || "");
      } else {
        Sentry.captureMessage(JSON.stringify(result));
      }
    }
  };

  firebaseLogin = async (idToken: string, accessToken: string) => {
    const credential = firebase.auth.GoogleAuthProvider.credential(
      idToken,
      accessToken
    );

    await firebase
      .auth()
      .signInAndRetrieveDataWithCredential(credential)
      .catch(error => {
        Sentry.captureMessage(JSON.stringify(error));
      });

    await this.setSession(true);
  };

  loggedIn = async () => {
    const idToken = await this.getIdToken();

    return Boolean(idToken);
  };

  logout = async () => {
    await firebase.auth().signOut();

    await AsyncStorage.removeItem("id_token");
    await AsyncStorage.removeItem("expiration");
    await AsyncStorage.removeItem("email");
  };

  setSession = async (refresh = false) => {
    const user = firebase.auth().currentUser;
    if (!user) {
      return;
    }

    if (user.email) {
      await AsyncStorage.setItem("email", user.email);
      await AsyncStorage.setItem("uid", user.uid);
      this.setState({
        email: user.email,
        uid: user.uid
      });
    }

    const idToken = await user.getIdToken(refresh);
    await AsyncStorage.setItem("id_token", idToken);
    await AsyncStorage.setItem(
      "expiration",
      String(new Date().getTime() + 60 * 60)
    );

    return idToken;
  };

  getIdToken = async () => {
    const idToken = await AsyncStorage.getItem("id_token");
    if (!idToken) {
      return null;
    }

    const expiration = await AsyncStorage.getItem("expiration");
    if (Number(expiration) > new Date().getTime()) {
      return idToken;
    }

    return this.setSession(true);
  };

  render() {
    return (
      <Provider
        value={{
          onGoogleLogin: this.onGoogleLogin,
          getIdToken: this.getIdToken,
          loggedIn: this.loggedIn,
          logout: this.logout,
          email: this.state.email,
          uid: this.state.uid
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
