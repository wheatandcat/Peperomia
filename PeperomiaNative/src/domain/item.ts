export type Item = {
  title: string;
  kind: string;
};

export type SelectItem = Item & {
  id: string | number;
};

export type UpdateItem = Item & {
  id: string | number;
};

export type DeleteItem = {
  id: string | number;
};
