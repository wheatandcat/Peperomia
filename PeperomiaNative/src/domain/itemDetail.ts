export type ItemDetail = {
  title: string;
  kind: string;
  memo: string;
  place: string;
  url: string;
  moveMinutes: number;
  priority: number;
};

export type SelectItemDetail = ItemDetail & {
  id?: string | number;
  itemId: string | number;
};

export type UpdateItemDetail = ItemDetail & {
  id: string | number;
  itemId: string | number;
};

export type DeleteItemDetail = {
  id: string | number;
};
