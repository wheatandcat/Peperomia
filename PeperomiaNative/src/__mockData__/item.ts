import { KIND_PARK, KIND_FISHING } from 'lib/getKind';
import { SelectItem } from 'domain/item';

export const itemsMockData: SelectItem[] = [
  {
    id: '1',
    title: '公園',
    kind: KIND_PARK,
  },
  {
    id: '2',
    title: '釣り',
    kind: KIND_FISHING,
  },
];

export const itemMockData: SelectItem = {
  id: '1',
  title: '公園',
  kind: KIND_PARK,
};
