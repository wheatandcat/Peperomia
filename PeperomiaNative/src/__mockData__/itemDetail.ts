import { ItemDetail } from 'domain/itemDetail';

export const itemDetailMockData: ItemDetail = {
  title: 'test',
  kind: 'park',
  memo: 'memo',
  place: 'place',
  url: 'url',
  moveMinutes: 30,
  priority: 1,
};

export const itemDetailsMockData: ItemDetail[] = [
  {
    title: 'test1',
    kind: 'park1',
    memo: 'memo1',
    place: 'place1',
    url: 'url1',
    moveMinutes: 30,
    priority: 1,
  },
  {
    title: 'test2',
    kind: 'park2',
    memo: 'memo2',
    place: 'place2',
    url: 'url2',
    moveMinutes: 45,
    priority: 2,
  },
];
