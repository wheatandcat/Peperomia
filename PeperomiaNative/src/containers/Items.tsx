import { SQLite } from "expo-sqlite";
import React, { createContext, Component } from "react";
import { db } from "../lib/db";
import { select as selectItems, Item } from "../lib/db/item";
import {
  selectByItemId as selectItemDetailByItemId,
  ItemDetail
} from "../lib/db/itemDetail";

const Context = createContext({});
const { Provider } = Context;

interface ItemAbout {
  itemId: number;
  about: string;
}

interface Props {}

interface State {
  loading: boolean;
  items: Item[];
  about: ItemAbout[];
}

export default class extends Component<Props, State> {
  state = {
    loading: true,
    items: [],
    about: []
  };

  componentDidMount() {
    this.getData();
  }

  getData = () => {
    db.transaction((tx: SQLite.Transaction) => {
      selectItems(tx, this.setItems);
    });
  };

  setItems = (data: any, error: any) => {
    if (error) {
      return;
    }
    this.setState({
      items: data,
      about: []
    });

    data.map((val: Item) => {
      db.transaction((tx: SQLite.Transaction) => {
        selectItemDetailByItemId(tx, String(val.id), this.setItemsDetail);
      });
    });
  };

  setItemsDetail = (data: any, error: any) => {
    if (error || !data || data.length === 0) {
      return;
    }

    const names = data.map((val: ItemDetail) => val.title).join("â†’");
    const itemId = data[0].itemId;
    const about = [
      ...this.state.about,
      {
        itemId: itemId,
        about: names
      }
    ];

    this.setState({
      about,
      loading: false
    });
  };

  render() {
    return (
      <Provider
        value={{
          items: this.state.items,
          about: this.state.about,
          refreshData: this.getData,
          itemsLoading: this.state.loading
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
