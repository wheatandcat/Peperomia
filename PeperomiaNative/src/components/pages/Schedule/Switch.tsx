import React, { Component } from 'react';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import {
  View,
  Share,
  AsyncStorage,
  Dimensions,
  Clipboard,
  Alert,
} from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet';
import {
  ActionSheetProps,
  connectActionSheet,
} from '@expo/react-native-action-sheet';
import Toast from 'react-native-root-toast';
import uuidv1 from 'uuid/v1';
import { Button } from 'react-native-elements';
import theme from '../../../config/theme';
import { Item } from '../../../domain/item';
import { SelectItemDetail } from '../../../domain/itemDetail';
import { updateItemDetail } from '../../../lib/itemDetail';
import { deleteItem, getItemByID } from '../../../lib/item';
import {
  save as saveFirestore,
  isShare,
  updateShare,
} from '../../../lib/firestore/plan';
import getShareText from '../../../lib/getShareText';
import {
  Consumer as ItemsConsumer,
  ContextProps,
} from '../../../containers/Items';
import SortableSchedule from '../SortableSchedule/Connected';
import Schedule from './Connected';
import HeaderLeft from './HeaderLeft';
import HeaderRight from './HeaderRight';

type State = Pick<Item, 'title'> & {
  item: Item;
  itemId: number;
  items: SelectItemDetail[];
  saveItems: SelectItemDetail[];
  mode: string;
};

type Props = ActionSheetProps & {
  navigation: NavigationScreenProp<NavigationRoute>;
};

type PlanProps = Props &
  State &
  Pick<ContextProps, 'refreshData'> &
  Pick<Switch, 'onShow' | 'onAdd' | 'onSort'>;

export class Switch extends Component<Props, State> {
  static navigationOptions = ({ navigation }: { navigation: any }) => {
    const { params = {} } = navigation.state;
    return {
      headerTitle: (
        <Button
          type="clear"
          title={params.title}
          onPress={params.onEditPlan}
          testID="ScheduleTitleUpdate"
          titleStyle={styles.headerTitle}
        />
      ),
      headerStyle: {
        backgroundColor: theme().mode.header.backgroundColor,
      },
      headerLeft: (
        <View style={styles.headerLeft}>
          <HeaderLeft
            mode={params.mode}
            onShow={params.onShow}
            navigation={navigation}
          />
        </View>
      ),
      headerRight: (
        <View style={styles.headerRight}>
          <HeaderRight
            mode={params.mode}
            onSave={params.onSave}
            onShare={() => params.onShare(params.title, params.items)}
            onOpenActionSheet={() =>
              params.onOpenActionSheet(
                params.itemId,
                params.title,
                params.items
              )
            }
          />
        </View>
      ),
    };
  };

  state = {
    item: {
      id: 0,
      title: '',
      kind: '',
      image: '',
    },
    itemId: 0,
    title: '',
    items: [],
    saveItems: [],
    mode: 'show',
  };

  async componentDidMount() {
    const itemId = this.props.navigation.getParam('itemId', '1');
    const item = await getItemByID<Item>(null, String(itemId));

    this.setState({
      item,
    });

    this.props.navigation.setParams({
      onAdd: this.onAdd,
      onShow: this.onShow,
      onSort: this.onSort,
      onSave: this.onSave,
      onShare: this.onShare,
      onEditPlan: this.onEditPlan,
      onOpenActionSheet: this.onOpenActionSheet,
      mode: 'show',
    });
  }

  onOpenActionSheet = async (
    itemId: string,
    title: string,
    items: SelectItemDetail[]
  ) => {
    const userID = await AsyncStorage.getItem('userID');
    if (userID) {
      const uuid = userID + itemId;
      const share = await isShare(uuid);
      if (share) {
        this.props.showActionSheetWithOptions(
          {
            options: [
              'リンクを取得する',
              'リンクを非公開にする',
              'その他',
              'キャンセル',
            ],
            destructiveButtonIndex: 1,
            cancelButtonIndex: 3,
          },
          buttonIndex => {
            if (buttonIndex === 0) {
              this.onCrateShareLink(items);
            } else if (buttonIndex === 1) {
              this.onCloseShareLink(uuid);
            } else if (buttonIndex === 2) {
              this.onShare(title, items);
            }
          }
        );
        return;
      }
    }

    this.props.showActionSheetWithOptions(
      {
        options: ['リンクを取得する', 'その他', 'キャンセル'],
        cancelButtonIndex: 2,
      },
      buttonIndex => {
        if (buttonIndex === 0) {
          Alert.alert(
            'この予定がWebで公開されます',
            'あとで非公開に変更することも可能です',
            [
              {
                text: 'キャンセル',
                style: 'cancel',
              },
              {
                text: '公開する',
                onPress: () => {
                  this.onCrateShareLink(items);
                },
              },
            ],
            { cancelable: false }
          );
        } else if (buttonIndex === 1) {
          this.onShare(title, items);
        }
      }
    );
  };

  onCrateShareLink = async (items: SelectItemDetail[]) => {
    if (!this.state.item.id) {
      return;
    }

    const userID = await AsyncStorage.getItem('userID');
    if (userID === null) {
      return;
    }

    const linkID = await saveFirestore(userID, this.state.item, items);
    if (!linkID) {
      Alert.alert('保存に失敗しました');
      return;
    }

    const shareHost = 'https://peperomia.info';
    console.log(`${shareHost}/${linkID}`);

    Clipboard.setString(`${shareHost}/${linkID}`);

    const { height } = Dimensions.get('window');

    const toast = Toast.show('リンクがコピーされました！', {
      duration: Toast.durations.LONG,
      position: height - 150,
      shadow: true,
      animation: true,
      hideOnPress: true,
      delay: 0,
    });

    // You can manually hide the Toast, or it will automatically disappear after a `duration` ms timeout.
    setTimeout(function() {
      Toast.hide(toast);
    }, 3000);
  };

  onCloseShareLink = async (doc: string) => {
    const result = await updateShare(doc, false);
    if (result) {
      const { height } = Dimensions.get('window');

      let toast = Toast.show('リンクを非公開にしました', {
        duration: Toast.durations.LONG,
        //textColor: "red",
        position: height - 150,
        shadow: true,
        animation: true,
        hideOnPress: true,
        delay: 0,
      });

      // You can manually hide the Toast, or it will automatically disappear after a `duration` ms timeout.
      setTimeout(function() {
        Toast.hide(toast);
      }, 3000);
    }
  };

  onAdd = (items: SelectItemDetail[]) => {
    const itemId = this.props.navigation.getParam('itemId', '1');
    this.props.navigation.navigate('AddScheduleDetail', {
      itemId,
      priority: items.length + 1,
      onSave: () => {
        this.props.navigation.navigate('Schedule', {
          itemId: itemId,
          refresh: uuidv1(),
        });
      },
    });
  };

  onShow = (): void => {
    this.setState({ mode: 'show' });

    this.props.navigation.setParams({
      mode: 'show',
    });
  };

  onSort = (items: SelectItemDetail[]): void => {
    this.setState({ mode: 'sort', items });

    this.props.navigation.setParams({
      mode: 'sort',
    });
  };

  onSave = () => {
    this.setState({
      items: this.state.saveItems,
    });
    this.onShow();
  };

  onShare = async (title: string, items: SelectItemDetail[]) => {
    try {
      const message = getShareText(items);

      const result = await Share.share({
        title,
        message,
      });

      if (result.action === Share.sharedAction) {
        if (result.activityType) {
          // shared with activity type of result.activityType
        } else {
          // shared
        }
      } else if (result.action === Share.dismissedAction) {
        // dismissed
      }
    } catch (error) {
      Alert.alert(error.message);
    }
  };

  onEditPlan = () => {
    this.props.navigation.navigate('EditPlan', {
      id: this.state.item.id,
      title: this.state.item.title,
      kind: this.state.item.kind,
      image: this.state.item.image,
    });
  };

  render() {
    return (
      <ItemsConsumer>
        {({ refreshData }: ContextProps) => (
          <Plan
            {...this.props}
            {...this.state}
            refreshData={refreshData}
            onShow={this.onShow}
            onAdd={this.onAdd}
            onSort={this.onSort}
          />
        )}
      </ItemsConsumer>
    );
  }
}

export class Plan extends Component<PlanProps> {
  onDelete = async () => {
    const itemId = this.props.navigation.getParam('itemId', '1');

    const ok = await deleteItem(null, { id: itemId });
    if (!ok) {
      Alert.alert('削除に失敗しました');
      return;
    }

    if (this.props.refreshData) {
      this.props.refreshData();
      this.props.navigation.goBack();
    }
  };

  onChangeItems = (data: SelectItemDetail[]) => {
    data.forEach(async (itemDetail, index) => {
      const v = {
        ...itemDetail,
        id: itemDetail.id || '',
        priority: index + 1,
      };

      const ok = await updateItemDetail(null, v);
      if (!ok) {
        Alert.alert('保存に失敗しました');
        return;
      }
    });
  };

  save = () => {
    if (this.props.refreshData) {
      this.props.refreshData();
    }
  };

  render() {
    if (this.props.mode === 'sort') {
      return (
        <SortableSchedule
          items={this.props.items}
          onChangeItems={this.onChangeItems}
        />
      );
    }

    return (
      <Schedule
        navigation={this.props.navigation}
        onAdd={this.props.onAdd}
        onSort={this.props.onSort}
        onDelete={this.onDelete}
      />
    );
  }
}

export default connectActionSheet(Switch);

const styles = EStyleSheet.create({
  headerTitle: {
    color: '$headerText',
    fontWeight: '600',
  },
  headerLeft: {
    left: 5,
  },
  headerRight: {
    right: 10,
  },
});
