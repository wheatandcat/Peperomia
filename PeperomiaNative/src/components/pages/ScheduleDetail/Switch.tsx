import React, { Component } from 'react';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import {
  Consumer as ItemsConsumer,
  ContextProps,
} from '../../../containers/Items';
import { ItemDetail as ItemDetailParam } from '../../../domain/itemDetail';
import ScheduleDetail from './Connected';
import EditScheduleDetail from '../EditScheduleDetail/Connected';

type State = ItemDetailParam & {
  scheduleDetailId: number;
  mode: string;
};

type Props = {
  navigation: NavigationScreenProp<NavigationRoute>;
};

type PlanProps = Props & Pick<ContextProps, 'refreshData'>;

export default class extends Component<Props> {
  render() {
    return (
      <ItemsConsumer>
        {({ refreshData }: ContextProps) => (
          <Plan {...this.props} refreshData={refreshData} />
        )}
      </ItemsConsumer>
    );
  }
}

class Plan extends Component<PlanProps, State> {
  static navigationOptions = { header: null };

  state = {
    title: '',
    memo: '',
    kind: '',
    url: '',
    place: '',
    moveMinutes: 0,
    priority: 0,
    scheduleDetailId: 0,
    mode: 'show',
  };

  onEdit = (
    title: string,
    kind: string,
    place: string,
    url: string,
    memo: string,
    moveMinutes: number,
    priority: number
  ): void => {
    const scheduleDetailId = this.props.navigation.getParam(
      'scheduleDetailId',
      '1'
    );

    this.setState({
      title,
      kind,
      memo,
      place,
      url,
      moveMinutes,
      priority,
      scheduleDetailId,
      mode: 'edit',
    });
  };

  onShow = (): void => {
    this.setState({ mode: 'show' });
  };

  render() {
    if (this.state.mode === 'edit') {
      return (
        <EditScheduleDetail
          id={this.state.scheduleDetailId}
          title={this.state.title}
          kind={this.state.kind}
          url={this.state.url}
          place={this.state.place}
          memo={this.state.memo}
          moveMinutes={this.state.moveMinutes}
          priority={this.state.priority}
          navigation={this.props.navigation}
          onShow={this.onShow}
        />
      );
    }

    return (
      <ScheduleDetail
        navigation={this.props.navigation}
        refreshData={this.props.refreshData}
        onEdit={this.onEdit}
      />
    );
  }
}
