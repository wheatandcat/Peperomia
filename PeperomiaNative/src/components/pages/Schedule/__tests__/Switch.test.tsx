import React from 'react';
import { Share, AsyncStorage } from 'react-native';
import { shallow, ShallowWrapper } from 'enzyme';
import * as itemDetail from '../../../../lib/itemDetail';
import * as item from '../../../../lib/item';
import * as plan from '../../../../lib/firestore/plan';
import * as Items from '../../../../containers/Items';
import * as Auth from '../../../../containers/Auth';
import Switch, { Connected, SwitchType } from '../Switch';

jest.mock('react-native-vector-icons', () => 'Icon');

describe('components/pages/Schedule/Switch.tsx', () => {
  describe('Switch', () => {
    let wrapper: ShallowWrapper;

    const navigate = jest.fn();
    const getParam = jest.fn();
    const setParams = jest.fn();

    const propsData: any = () => ({
      navigation: {
        navigate,
        getParam,
        setParams,
        goBack: jest.fn(),
      },
    });

    beforeEach(() => {
      jest.spyOn(Auth, 'useAuth').mockImplementation(
        () =>
          ({
            uid: '1',
          } as any)
      );

      jest.spyOn(Items, 'useItems').mockImplementation(
        () =>
          ({
            refreshData: jest.fn(),
          } as any)
      );

      wrapper = shallow(<Switch {...propsData()} />);
    });

    it('正常に表示されている', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Connected', () => {
    let wrapper: ShallowWrapper<SwitchType>;

    const navigate = jest.fn();
    const getParam = jest.fn();
    const setParams = jest.fn();
    const showActionSheetWithOptions = jest.fn();

    const propsData: any = () => ({
      uid: '1',
      refreshData: jest.fn(),
      showActionSheetWithOptions,
      navigation: {
        navigate,
        getParam,
        setParams,
        goBack: jest.fn(),
      },
    });

    beforeEach(() => {
      jest.spyOn(item, 'getItemByID').mockImplementation(
        () =>
          ({
            id: '1',
            title: 'test',
            kind: 'test',
          } as any)
      );

      jest.spyOn(React, 'useEffect').mockImplementation(async f => await f());

      wrapper = shallow(<Connected {...propsData()} />);
    });

    it('正常に表示されている', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('onAdd', () => {
      wrapper.props().onAdd([]);
    });

    it('onDelete', async () => {
      jest.spyOn(item, 'deleteItem').mockImplementation(() => true as any);

      await wrapper.props().onDelete();
    });

    it('onShow', () => {
      setParams.mock.calls[0][0].onShow();
      expect(setParams).toHaveBeenCalledWith({
        mode: 'show',
      });
    });

    it('onSave => 成功', async () => {
      jest
        .spyOn(itemDetail, 'updateItemDetail')
        .mockImplementation(() => true as any);

      await setParams.mock.calls[0][0].onSave();
    });

    it('onSave => 失敗', async () => {
      jest
        .spyOn(itemDetail, 'updateItemDetail')
        .mockImplementation(() => false as any);

      await setParams.mock.calls[0][0].onSave();
    });

    it('onEditPlan', () => {
      setParams.mock.calls[4][0].onEditPlan({
        id: '1',
        title: 'test',
        kind: 'test',
      });
    });

    it('onSort', async () => {
      jest
        .spyOn(itemDetail, 'getItemDetails')
        .mockImplementation(() => [] as any);

      await wrapper.props().onSort();
      expect(setParams).toHaveBeenCalledWith({
        mode: 'sort',
      });
    });

    it('onShare', async () => {
      jest
        .spyOn(itemDetail, 'getItemDetails')
        .mockImplementation(() => [] as any);

      jest.spyOn(Share, 'share').mockImplementation(() => ({} as any));

      await setParams.mock.calls[0][0].onShare('test', []);
    });

    it('onOpenActionSheet => userID無し', async () => {
      jest.spyOn(plan, 'isShare').mockImplementation(() => true as any);

      await setParams.mock.calls[0][0].onOpenActionSheet('1', 'test', []);

      showActionSheetWithOptions.mock.calls[0][1](0);
      showActionSheetWithOptions.mock.calls[0][1](1);
    });

    it('onOpenActionSheet => userID => 未公開', async () => {
      jest.spyOn(plan, 'isShare').mockImplementation(() => false as any);
      jest
        .spyOn(AsyncStorage, 'getItem')
        .mockImplementation(() => 'test' as any);

      await setParams.mock.calls[0][0].onOpenActionSheet('1', 'test', []);
    });

    it('onOpenActionSheet => userID => 公開', async () => {
      jest.spyOn(plan, 'isShare').mockImplementation(() => true as any);
      jest
        .spyOn(AsyncStorage, 'getItem')
        .mockImplementation(() => 'test' as any);

      await setParams.mock.calls[0][0].onOpenActionSheet('1', 'test', []);

      showActionSheetWithOptions.mock.calls[2][1](0);
      showActionSheetWithOptions.mock.calls[2][1](1);
      showActionSheetWithOptions.mock.calls[2][1](2);
    });
  });
});
