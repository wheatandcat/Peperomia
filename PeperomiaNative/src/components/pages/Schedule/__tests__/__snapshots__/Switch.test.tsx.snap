// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/Schedule/Switch.tsx Connected 正常に表示されている 1`] = `
<Memo()
  itemDetails={Array []}
  navigation={
    Object {
      "getParam": [MockFunction],
      "goBack": [MockFunction],
      "navigate": [MockFunction],
      "setParams": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "mode": "show",
              "onOpenActionSheet": [Function],
              "onSave": [Function],
              "onShare": [Function],
              "onShow": [Function],
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
  onAdd={[Function]}
  onDelete={[Function]}
  onSort={[Function]}
  route={
    Object {
      "params": Object {
        "itemId": 1,
        "title": "test",
      },
    }
  }
/>
`;

exports[`components/pages/Schedule/Switch.tsx Switch 正常に表示されている 1`] = `
<Memo()
  navigation={
    Object {
      "getParam": [MockFunction],
      "goBack": [MockFunction],
      "navigate": [MockFunction],
      "setParams": [MockFunction],
    }
  }
  refreshData={[MockFunction]}
  route={
    Object {
      "params": Object {
        "itemId": 1,
        "title": "test",
      },
    }
  }
  showActionSheetWithOptions={[Function]}
  uid="1"
/>
`;
